Sources/GRPCCore/Server.swift

```diff
@@ -0,0 +1,428 @@
+/// can handle the RPC.
+///
+/// A ``Server`` may listen with multiple transports (for example, HTTP/2 and in-process) and route
+/// requests from each transport to the same service instance. You can also use "interceptors",
+/// to implement cross-cutting logic which apply to all accepted RPCs. Example uses of interceptors
+/// include request filtering, authentication, and logging. Once requests have been intercepted
+/// they are passed to a handler which in turn returns a response to send back to the client.
+///
+/// ## Creating and configuring a server
+///
+/// The following example demonstrates how to create and configure a server.
+///
+/// ```swift
+/// var server = Server()
+///
+/// // Create and add an in-process transport.
+/// let inProcessTransport = InProcessServerTransport()
+/// server.transports.add(inProcessTransport)
+///
+/// // Create and register the 'Greeter' and 'Echo' services.
+/// server.services.register(GreeterService())
+/// server.services.register(EchoService())
+///
+/// // Create and add some interceptors.
+/// server.interceptors.add(StatsRecordingServerInterceptors())
+/// ```
+///
+/// ## Starting and stopping the server
+///
+/// Once you have configured the server call ``run()`` to start it. Calling ``run()`` starts each
+/// of the servers transports. A ``ServerError`` is thrown if any of the transports can't be
+/// started.
+///
+/// ```swift
+/// // Start running the server.
+/// try await server.run()
+/// ```
+///
+/// The ``run()`` method won't return until the server has finished handling all requests. You can
+/// signal to the server that it should stop accepting new requests by calling ``stopListening()``.
+/// This allows the server to drain existing requests gracefully. To stop the server more abruptly
+/// you can cancel the task running your server. If your services require additional resources that
+/// need their lifecycles managed you should consider using [Swift Service
+/// Lifecycle](https://github.com/swift-server/swift-service-lifecycle).
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public struct Server: Sendable {
+  typealias Stream = RPCStream<ServerTransport.Inbound, ServerTransport.Outbound>
+
+  /// A collection of ``ServerTransport`` implementations that the server uses to listen
+  /// for new requests.
+  public var transports: Transports
+
+  /// The services registered which the server is serving.
+  public var services: Services
+
+  /// A collection of ``ServerInterceptor`` implementations which are applied to all accepted
+  /// RPCs.
+  ///
+  /// RPCs are intercepted in the order that interceptors are added. That is, a request received
+  /// from the client will first be intercepted by the first added interceptor followed by the
+  /// second, and so on.
+  public var interceptors: Interceptors
+
+  /// The state of the server.
+  private enum State {
```

Comment:
Can we move the transports & interceptors into the `State` so we release them as early as possible?